contracts/src/universalResolver/RegistryUtils.sol

```diff
@@ -0,0 +1,132 @@
+            next
+        );
+        // if there was a parent registry...
+        if (address(registry) != address(0)) {
+            string memory label = readLabel(name, offset);
+            // remember the resolver (if it exists)
+            address res = registry.getResolver(label);
+            if (res != address(0)) {
+                resolver = res;
+                resolverOffset = offset;
+            }
+            registry = registry.getSubregistry(label);
+        }
+        // update namehash
+        node = keccak256(abi.encode(node, labelHash));
+    }
+
+    /// @notice Find the exact registry for `name[offset:]`.
+    /// @param rootRegistry The root ENS registry.
+    /// @param name The DNS-encoded name to search.
+    /// @return registry The exact registry or null if not found.
+    function findExactRegistry(
+        IRegistry rootRegistry,
+        bytes memory name,
+        uint256 offset
+    ) internal view returns (IRegistry registry) {
+        (bytes32 labelHash, uint256 next) = NameCoder.readLabel(name, offset);
+        if (labelHash == bytes32(0)) {
+            return rootRegistry;
+        }
+        IRegistry parent = findExactRegistry(rootRegistry, name, next);
+        if (address(parent) != address(0)) {
+            string memory label = readLabel(name, offset);
+            registry = parent.getSubregistry(label);
+        }
+    }
+
+    /// @notice Find the parent registry for `name[offset:]`.
+    /// @param rootRegistry The root ENS registry.
+    /// @param name The DNS-encoded name to search.
+    /// @return parentRegistry The parent registry or null if not found.
+    function findParentRegistry(
+        IRegistry rootRegistry,
+        bytes calldata name,
+        uint256 offset
+    ) external view returns (IRegistry parentRegistry) {
+        (bytes32 labelHash, uint256 next) = NameCoder.readLabel(name, offset);
+        if (labelHash != bytes32(0)) {
+            parentRegistry = findExactRegistry(rootRegistry, name, next);
+        }
+    }
+
+    /// @dev Read label at offset from a DNS-encoded name.
+    ///      eg. `readLabel("\x03abc\x00", 0) = "abc"`.
+    /// @param name The name.
+    /// @param offset The offset into `name`
+    /// @return label The label.
+    function readLabel(
+        bytes memory name,
+        uint256 offset
+    ) internal pure returns (string memory label) {
+        (uint8 size, ) = nextLabel(name, offset);
+        label = new string(size);
+        assembly {
+            mcopy(add(label, 32), add(add(name, 33), offset), size)
+        }
+    }
+
+    // ********************************************************************************
+    // I'd like to move the following code here:
+    // https://github.com/ensdomains/ens-contracts/pull/459
+    // https://github.com/ensdomains/ens-contracts/blob/fix/namecoder-iter/contracts/utils/NameCoder.sol#L42
+    // ********************************************************************************
```

Comment:
If this is something you can't do right now, label it as a TODO with a specific condition when it can happen.